### PR TITLE
Rework exception messages

### PR DESCRIPTION
**Pull request by sondrelg**
Exceptions are reworked to not contain any formatting logic. Validators and parts of the schema tester are reworked to provide more informative error messages.

Ended up with a few different formats.

## Simplest format

The following validators:
- validate_min_properties
- validate_max_properties
- validate_max_items
- validate_min_items
- validate_max_length
- validate_min_length
- validate_unique_items
- validate_minimum
- validate_exclusive_minimum
- validate_maximum
- validate_exclusive_maximum
- validate_multiple_of
- validate_pattern

as well as `anyOf`, `oneOf` errors return simple messages without any special formatting, so will end up looking like this:

```python
E   openapi_tester.exceptions.DocumentationError: {message}
E   
E   Reference: {reference}
```

### Enum validator

The enum validator now looks like this

```python
E   openapi_tester.exceptions.DocumentationError: 
E   
E   Expected: a member of the enum ["Cat", "Dog"]
E
E   Received: "Turtle"
E
E   Reference: {reference}
```

### Format validators

The format validators now look like this

```python
E   openapi_tester.exceptions.DocumentationError: 
E   
E   Expected: a "byte" formatted value
E
E   Received: "not byte"
E
E   Reference: {reference}
```

### Type validators

Type validators now look like

```python
E   openapi_tester.exceptions.DocumentationError: 
E   
E   Expected: a "string" type value
E
E   Received: 1
E
E   Reference: {reference}
```

**Note: We do no longer show a faked version of the schema at all.**

## Changes summarized

- Removed all `.` from error messages
- Split up consolidated validators - one validator for each error message
- Removed `DocumentationError` formatting and moved message formatting to validators
- Subclassed case exception as `DocumentationError`

### CODE REVIEW

File: openapi_tester/constants.py [hunk 1]
```diff
@@ -14,37 +14,37 @@
...
 VALIDATE_MAXIMUM_ERROR = "The response value {data} exceeds the maximum allowed value of {maximum}"
-VALIDATE_MIN_LENGTH_ERROR = "The length of {data} exceeds the minimum allowed length of {min_length}"
-VALIDATE_MAX_LENGTH_ERROR = "The length of {data} exceeds the maximum allowed length of {max_length}"
+VALIDATE_MIN_LENGTH_ERROR = 'The length of "{data}" is below the minimum required length of {min_length}'
+VALIDATE_MAX_LENGTH_ERROR = 'The length of "{data}" exceeds the maximum allowed length of {max_length}'
 VALIDATE_MIN_ARRAY_LENGTH_ERROR = "The length of the array {data} is below the minimum required length of {min_length}"
 VALIDATE_MAX_ARRAY_LENGTH_ERROR = "The length of the array {data} exceeds the maximum allowed length of {max_length}"
 VALIDATE_MINIMUM_NUMBER_OF_PROPERTIES_ERROR = (
-    "The number of properties in {data} is below the minimum number required, {min_length}"
+    "The number of properties in {data} is below {min_length} which is the minimum number of properties required"
```
Comment (Goldziher) [inline openapi_tester/constants.py:30]: Too verbose. Is below the specified minimum of ... thats it

File: openapi_tester/constants.py [hunk 2]
```diff
@@ -14,37 +14,37 @@
...
 )
 VALIDATE_MAXIMUM_NUMBER_OF_PROPERTIES_ERROR = (
-    "The number of properties in {data} exceeds the maximum number allowed, {max_length}"
+    "The number of properties in {data} exceeds {max_length} which is the maximum number of properties allowed"
```
Comment (Goldziher) [inline openapi_tester/constants.py:33]: Ibid.

**Review by Goldziher — APPROVED** [review]

File: openapi_tester/constants.py [hunk 3]
```diff
@@ -14,37 +14,37 @@
 PARAMETER_CAPTURE_REGEX = re.compile(r"({[\w]+})")
 
 # Validation errors
-VALIDATE_FORMAT_ERROR = "Mismatched values, expected a value with the format {expected} but received {received}."
-VALIDATE_PATTERN_ERROR = "String '{data}' does not validate using the specified pattern: {pattern}"
+VALIDATE_FORMAT_ERROR = 'Expected: {article} "{format}" formatted value, like {example}\n\nReceived: {received}'
```
Comment (Goldziher) [inline openapi_tester/constants.py:17]:
```suggestion
VALIDATE_FORMAT_ERROR = 'Expected: {article} "{format}" formatted value\n\nReceived: {received}'
```

File: openapi_tester/constants.py [hunk 4]
```diff
@@ -14,37 +14,37 @@
...
+VALIDATE_PATTERN_ERROR = 'The string "{data}" does not validate using the specified pattern: {pattern}'
```
Comment (Goldziher) [inline openapi_tester/constants.py:18]:
```suggestion
VALIDATE_PATTERN_ERROR = 'The string "{data}" does not match the specified pattern: {pattern}'
```

File: openapi_tester/constants.py [hunk 5]
```diff
@@ -14,37 +14,37 @@
...
 INVALID_PATTERN_ERROR = "String pattern is not valid regex: {pattern}"
-VALIDATE_ENUM_ERROR = "Mismatched values, expected a member of the enum {enum} but received {received}."
-VALIDATE_TYPE_ERROR = "Mismatched types, expected {expected} but received {received}."
+VALIDATE_ENUM_ERROR = "Expected: a valid enum member, like {expected}\n\nReceived: {received}"
```
Comment (Goldziher) [inline openapi_tester/constants.py:20]:
```suggestion
VALIDATE_ENUM_ERROR = "Expected: a member of the enum {enum}\n\nReceived: {received}"
```

File: openapi_tester/constants.py [hunk 6]
```diff
@@ -14,37 +14,37 @@
...
+VALIDATE_TYPE_ERROR = 'Expected: {article} "{type}" type value\n\nReceived: {received}'
 VALIDATE_MULTIPLE_OF_ERROR = "The response value {data} should be a multiple of {multiple}"
-VALIDATE_MINIMUM_ERROR = "The response value {data} exceeds the minimum allowed value of {minimum}"
+VALIDATE_MINIMUM_ERROR = "The response value {data} is below the minimum required value of {minimum}"
```
Comment (Goldziher) [inline openapi_tester/constants.py:23]:
```suggestion
VALIDATE_MINIMUM_ERROR = "The response value {data} is lower than the specified minimum of {minimum}"
```

File: openapi_tester/constants.py [hunk 7]
```diff
@@ -14,37 +14,37 @@
...
 VALIDATE_MAXIMUM_ERROR = "The response value {data} exceeds the maximum allowed value of {maximum}"
-VALIDATE_MIN_LENGTH_ERROR = "The length of {data} exceeds the minimum allowed length of {min_length}"
-VALIDATE_MAX_LENGTH_ERROR = "The length of {data} exceeds the maximum allowed length of {max_length}"
+VALIDATE_MIN_LENGTH_ERROR = 'The length of "{data}" is below the minimum required length of {min_length}'
```
Comment (Goldziher) [inline openapi_tester/constants.py:25]:
```suggestion
VALIDATE_MIN_LENGTH_ERROR = 'The length of "{data}" is shorter than the specified minimum length of {min_length}'
```

File: openapi_tester/constants.py [hunk 8]
```diff
@@ -14,37 +14,37 @@
...
+VALIDATE_MAX_LENGTH_ERROR = 'The length of "{data}" exceeds the maximum allowed length of {max_length}'
```
Comment (Goldziher) [inline openapi_tester/constants.py:26]:
```suggestion
VALIDATE_MAX_LENGTH_ERROR = 'The length of "{data}" exceeds the specified maximum length of {max_length}'
```

File: openapi_tester/constants.py [hunk 9]
```diff
@@ -14,37 +14,37 @@
...
 VALIDATE_MIN_ARRAY_LENGTH_ERROR = "The length of the array {data} is below the minimum required length of {min_length}"
```
Comment (Goldziher) [inline openapi_tester/constants.py:27]:
```suggestion
VALIDATE_MIN_ARRAY_LENGTH_ERROR = "The length of the array {data} is shorter than the specified minimum length of {min_length}"
```

File: openapi_tester/constants.py [hunk 10]
```diff
@@ -14,37 +14,37 @@
...
 VALIDATE_MAX_ARRAY_LENGTH_ERROR = "The length of the array {data} exceeds the maximum allowed length of {max_length}"
```
Comment (Goldziher) [inline openapi_tester/constants.py:28]:
```suggestion
VALIDATE_MAX_ARRAY_LENGTH_ERROR = "The length of the array {data} exceeds the specified maximum length of {max_length}"
```

File: openapi_tester/constants.py (hunk 1 above)
Comment (Goldziher) [inline openapi_tester/constants.py:30]:
```suggestion
    "The number of properties in {data} is fewer than the specified minimum number of properties of {min_length}"
```

File: openapi_tester/constants.py (hunk 2 above)
Comment (Goldziher) [inline openapi_tester/constants.py:33]:
```suggestion
    "The number of properties in {data} exceeds the specified maximum number of properties of {max_length}"
```

File: openapi_tester/constants.py [hunk 11]
```diff
@@ -14,37 +14,37 @@
...
+    "The number of properties in {data} is below {min_length} which is the minimum number of properties required"
 )
 VALIDATE_MAXIMUM_NUMBER_OF_PROPERTIES_ERROR = (
-    "The number of properties in {data} exceeds the maximum number allowed, {max_length}"
+    "The number of properties in {data} exceeds {max_length} which is the maximum number of properties allowed"
 )
-VALIDATE_UNIQUE_ITEMS_ERROR = "The schema specifies that the array must have unique items, but items are not unique."
-VALIDATE_NONE_ERROR = "Mismatched content. Expected {expected} but received NoneType"
-VALIDATE_MISSING_RESPONSE_KEY_ERROR = "The following property is missing from the tested data: {missing_key}."
+VALIDATE_UNIQUE_ITEMS_ERROR = "The array {data} must only contain unique items"
```
Comment (Goldziher) [inline openapi_tester/constants.py:35]:
```suggestion
VALIDATE_UNIQUE_ITEMS_ERROR = "The array {data} must contain unique items only"
```

File: openapi_tester/constants.py [hunk 12]
```diff
@@ -14,37 +14,37 @@
...
+VALIDATE_NONE_ERROR = "Received null value for a non-nullable schema object"
```
Comment (Goldziher) [inline openapi_tester/constants.py:36]:
```suggestion
VALIDATE_NONE_ERROR = "Received a null value for a non-nullable schema object"
```

File: openapi_tester/constants.py [hunk 13]
```diff
@@ -14,37 +14,37 @@
...
+VALIDATE_MISSING_RESPONSE_KEY_ERROR = 'The following property is missing from your response: "{missing_key}"'
```
Comment (Goldziher) [inline openapi_tester/constants.py:37]:
```suggestion
VALIDATE_MISSING_RESPONSE_KEY_ERROR = 'The following property is missing in the response data: "{missing_key}"'
```

File: openapi_tester/constants.py [hunk 14]
```diff
@@ -14,37 +14,37 @@
...
 VALIDATE_MISSING_PROPERTY_KEY_ERROR = (
     "The following key was found in your required properties, but is missing from properties: {missing_key}"
```
Comment (Goldziher) [inline openapi_tester/constants.py:39]: I'd drop all references to "you".
```suggestion
    "The following key is specified as required but is missing from the schema's properties: {missing_key}"
```

File: openapi_tester/validators.py [hunk 15]
```diff
@@ -74,24 +73,64 @@ def wrapped(value: Any):
...
+            received=f'"{data}"' if isinstance(data, str) else data,
         )
     return None
 
 
 def validate_format(schema_section: Dict[str, Any], data: Any) -> Optional[str]:
     schema_format: str = schema_section.get("format", "")
     if schema_format in VALIDATOR_MAP and not VALIDATOR_MAP[schema_format](data):
-        return VALIDATE_FORMAT_ERROR.format(expected=schema_section["format"], received=str(data))
+        example_values = {
```
Comment (Goldziher) [inline openapi_tester/validators.py:88]: I'd completely remove the example.

The only think that might be needed here is a map, but really I think we can also do without the article :),

expected: "email" formatted value
received: {value}

This is simple, low maintainance and we lose no cohesion in doing this.

File: openapi_tester/validators.py [hunk 16]
```diff
@@ -74,24 +73,64 @@ def wrapped(value: Any):
...
+        return VALIDATE_FORMAT_ERROR.format(
+            format=schema_format,
+            article=example_values[schema_format]["article"],  # type: ignore
+            example=example_values[schema_format]["value"],  # type: ignore
+            received=f'"{data}"',
+        )
     return None
 
 
+def pretty_print_list(_list: list) -> str:
```
Comment (Goldziher) [inline openapi_tester/validators.py:111]: please remove this method completely. 

You can just use the excellent built-in library called pprint: https://docs.python.org/3/library/pprint.html

**Review by Goldziher — CHANGES_REQUESTED** [review]

File: openapi_tester/constants.py (hunk 14 above)
Comment (sondrelg) [reply openapi_tester/constants.py:39]: Wouldn't it be `missing from your response's properties` here?

File: openapi_tester/validators.py (hunk 15 above)
Comment (sondrelg) [reply openapi_tester/validators.py:88]: Yeah lets do this 👍

File: openapi_tester/validators.py (hunk 16 above)
Comment (sondrelg) [reply openapi_tester/validators.py:111]: Yep removing it as we changed the enum string anyways.

File: openapi_tester/constants.py (hunk 14 above)
Comment (Goldziher) [reply openapi_tester/constants.py:39]: I guess so 😃, just see where its raised.

Comment (sondrelg) [reply openapi_tester/constants.py:39]: This constant actually no longer seems to be used anywhere

Comment (sondrelg) [reply openapi_tester/constants.py]: Deleted it :)

**Review by Goldziher — APPROVED** [review]

**Pull request merged**